src/com/amazon/corretto/crypto/provider/AesCbcSpi.java

```diff
@@ -0,0 +1,685 @@
+    }
+
+    // If padding is enabled and decrypting, the largest output size is during doFinal
+    return (int) all;
+  }
+
+  @Override
+  protected byte[] engineGetIV() {
+    return iv == null ? null : iv.clone();
+  }
+
+  @Override
+  protected AlgorithmParameters engineGetParameters() {
+    try {
+      AlgorithmParameters parameters = AlgorithmParameters.getInstance("AES");
+      byte[] ivForParams = iv;
+      if (ivForParams == null) {
+        // We aren't initialized, so we return default and random values
+        ivForParams = new byte[BLOCK_SIZE_IN_BYTES];
+        new LibCryptoRng().nextBytes(ivForParams);
+      }
+      parameters.init(new IvParameterSpec(ivForParams));
+      return parameters;
+    } catch (final InvalidParameterSpecException | NoSuchAlgorithmException e) {
+      throw new Error("Unexpected error", e);
+    }
+  }
+
+  @Override
+  protected void engineInit(final int opmode, final Key key, final SecureRandom random)
+      throws InvalidKeyException {
+    if (opmode != Cipher.ENCRYPT_MODE && opmode != Cipher.WRAP_MODE) {
+      throw new InvalidKeyException("IV required for decrypt");
+    }
+
+    final byte[] iv = new byte[BLOCK_SIZE_IN_BYTES];
+    random.nextBytes(iv);
+
+    try {
+      engineInit(opmode, key, new IvParameterSpec(iv), null);
+    } catch (final InvalidAlgorithmParameterException e) {
+      throw new RuntimeCryptoException(e);
+    }
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode, final Key key, final AlgorithmParameters params, final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    try {
+      engineInit(opmode, key, params.getParameterSpec(IvParameterSpec.class), null);
+    } catch (final InvalidParameterSpecException e) {
+      throw new InvalidAlgorithmParameterException(e);
+    }
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode,
+      final Key key,
+      final AlgorithmParameterSpec params,
+      final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    final int opMode = checkOperation(opmode);
+    final byte[] iv = checkAesCbcIv(params);
+    final byte[] keyBytes = checkAesKey(key);
+
+    // All checks passes, so we update the state:
+    this.cipherState = CipherState.INITIALIZED;
+    this.opMode = opMode;
+    this.iv = iv;
+    this.key = keyBytes;
+    this.unprocessedInput = 0;
+  }
+
+  private static int checkOperation(final int opMode) throws InvalidParameterException {
+    return ((opMode == Cipher.ENCRYPT_MODE) || (opMode == Cipher.WRAP_MODE)) ? ENC_MODE : DEC_MODE;
+  }
+
+  private static byte[] checkAesCbcIv(final AlgorithmParameterSpec params)
+      throws InvalidAlgorithmParameterException {
+    if (!(params instanceof IvParameterSpec)) {
+      throw new InvalidAlgorithmParameterException(
+          "I don't know how to handle a " + params.getClass());
```

Comment:
Let's not give ACCP a first person voice.

Reply:
Sure.